### PR TITLE
Improve process creation demos in the Compute lecture

### PR DESCRIPTION
**Pull request by teodutu**
The lecture was missing a simple `fork()` demo. It jumped straight to `fork()` + `exec()`. A more granular introduction feels needed.

### CODE REVIEW

File: content/chapters/compute/lecture/demo/create-process/multiple_forks.c [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "utils.h"
+
+int main(void)
+{
+	pid_t pid;
+	int rc;
+	int exit_code;
+
+	fork();
```
Comment (razvand) [inline content/chapters/compute/lecture/demo/create-process/multiple_forks.c:19]: This doesn't help show the process hierarchy which is what you want to have with this demo.

I suggest you show the PID of the current process, before / after each `fork()`. And also, after all forks, do a `wait_for_input()` and then do a `pstree -p` or `ps -H` to show the process hierarchy.

**Review by razvand — CHANGES_REQUESTED** [review]
Hi, @teodutu , thanks. Please see my comment. And please add a small `README.md` file in each demo file showing how to run it and what you should observe. It's difficult when you teach: just see the code and no other info on building, running, testing and showing.

**Comment by teodutu** [comment]
Done @razvand. Please take another look. You too, @Alex-deVis.

File: content/chapters/compute/lecture/demo/create-process/README.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+## Running
+
+### `system.c`
+
+This is the most straightforward way to create a process in C.
+Just specify the path of the new process, together with all its arguments as a string.
+
+### `posix_spawn.c`
+
+`posix_spawn()` is similar to `system()`, but it's more granular.
```
Comment (Alex-deVis) [inline content/chapters/compute/lecture/demo/create-process/README.md:19]:
```suggestion
`posix_spawn()` is similar to `system()`, but more granular.
```

File: content/chapters/compute/lecture/demo/create-process/README.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+Now we can specify the `argv` and `envp` received by the spawned process's `main()` function.
+
+### `create_process.c`
+
+The windows **syscall** `CreateProcess()` is quie similar to the Linxu **library function** `posix_spawn()`.
```
Comment (Alex-deVis) [inline content/chapters/compute/lecture/demo/create-process/README.md:24]:
```suggestion
The windows **syscall** `CreateProcess()` is quie similar to the Linux **library function** `posix_spawn()`.
```

**Comment by teodutu** [comment]
@razvand In case you think that having to press Enter for each process spawned by `fork()` is a problem, @Alex-deVis made a fix here: https://github.com/teodutu/operating-systems/pull/5. Give it a look and tell us if you find a better solution or if you also think the way it is now is ok.

File: content/chapters/compute/lecture/demo/create-process/README.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+Run the code and see how `fork()` returns **twice**.
+It creates another process and then returns a different value to each of the two processes:
+- it returns 0 to the child process
+- it returns the PID of the child process to the parent
+
+### `multiple_forks.c`
+
+Unlike the example in `fork.c`, this one calls `fork()` twice more.
+Can you tell how many processes there will be in the end?
```
Comment (Alex-deVis) [inline content/chapters/compute/lecture/demo/create-process/README.md:42]:
```suggestion
Can you tell how many processes will be in the end?
```

File: content/chapters/compute/lecture/demo/create-process/README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+Run the code and analyse the process hierarchy using the following output:
+
+```
+After first fork: PID = 40700; PPID = 9549
+After second fork: PID = 40700; PPID = 9549
+After first fork: PID = 40701; PPID = 40700
+After second fork: PID = 40702; PPID = 40700
+After second fork: PID = 40701; PPID = 40700
+After second fork: PID = 40704; PPID = 40701
```
Comment (Alex-deVis) [inline content/chapters/compute/lecture/demo/create-process/README.md:52]:
```suggestion
After first fork: PID = 33501; PPID = 16706
After first fork: PID = 33502; PPID = 33501

After second fork: PID = 33501; PPID = 16706
After second fork: PID = 33502; PPID = 33501
After second fork: PID = 33520; PPID = 33501
After second fork: PID = 33521; PPID = 33502
```

File: content/chapters/compute/lecture/demo/create-process/README.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+Investigate the process hierarchy after each `fork()`.
+In the end, it should look like this:
+
+```console
+student@os:~/.../lab/support/create-process$ $ pstree -p $(pidof multiple_forks)
+multiple_forks(41201)─┬─multiple_forks(41202)─┬─multiple_forks(41230)───multiple_forks(41649)
+                      │                       └─multiple_forks(41644)
+                      ├─multiple_forks(41225)───multiple_forks(41641)
+                      └─multiple_forks(41247)
```
Comment (Alex-deVis) [inline content/chapters/compute/lecture/demo/create-process/README.md:66]:
```suggestion
student@os:~/.../lab/support/create-process$ $ pstree -p $(pidof multiple_forks)
multiple_forks(33501)─┬─multiple_forks(33502)─┬─multiple_forks(33521)───multiple_forks(33533)
                      │                       └─multiple_forks(33532)
                      ├─multiple_forks(33520)───multiple_forks(33534)
                      └─multiple_forks(33531)
```

File: content/chapters/compute/lecture/demo/create-process/README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+After first fork: PID = 40700; PPID = 9549
+After second fork: PID = 40700; PPID = 9549
+After first fork: PID = 40701; PPID = 40700
+After second fork: PID = 40702; PPID = 40700
+After second fork: PID = 40701; PPID = 40700
+After second fork: PID = 40704; PPID = 40701
+```
+
+Note that after they're created, the processes act independently.
+Therefore, you will to press `Enter` for each newly created process (as prompted by `wait_for_input()`.
```
Comment (Alex-deVis) [inline content/chapters/compute/lecture/demo/create-process/README.md:56]:
```suggestion
Note that after they're created, the processes act independently.
```
Remove this line if you go with my approach.

**Review by Alex-deVis — CHANGES_REQUESTED** [review]
Updated multiple forks output in README.

File: content/chapters/compute/lecture/demo/create-process/README.md (hunk 6 above)
Comment (teodutu) [reply content/chapters/compute/lecture/demo/create-process/README.md:66]: I don't get it. The only difference lies in the PIDs which are always going to be different.

File: content/chapters/compute/lecture/demo/create-process/README.md (hunk 5 above)
Comment (teodutu) [reply content/chapters/compute/lecture/demo/create-process/README.md:52]: I see your point, but the fact that the prints appear intermingled is part of the demo IMO. I wouldn't change this.

File: content/chapters/compute/lecture/demo/create-process/README.md (hunk 6 above)
Comment (Alex-deVis) [reply content/chapters/compute/lecture/demo/create-process/README.md:66]: Only made them consistent with the previous output.

File: content/chapters/compute/lecture/demo/create-process/README.md (hunk 5 above)
Comment (Alex-deVis) [reply content/chapters/compute/lecture/demo/create-process/README.md:52]: Agreed.

File: content/chapters/compute/lecture/demo/create-process/README.md (hunk 4 above)
Comment (teodutu) [reply content/chapters/compute/lecture/demo/create-process/README.md:42]: My wording isn't incorrect. The answer would be "In the end, there will be X processes.", so the question, taken separately, would be "How many processes will there be in the end?". Therefore, when this sentence becomes the object of the first sentence in the phrase, the topic is swapped back: "how many processes there will be".

File: content/chapters/compute/lecture/demo/create-process/README.md (hunk 6 above)
Comment (teodutu) [reply content/chapters/compute/lecture/demo/create-process/README.md:66]: They're going to differ when students run the code anyway, so consistency here is not important.

File: content/chapters/compute/lecture/demo/create-process/README.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+Investigate the process hierarchy after each `fork()`.
+In the end, it should look like this:
+
+```console
+student@os:~/.../lab/support/create-process$ $ pstree -p $(pidof multiple_forks)
```
Comment (razvand) [inline content/chapters/compute/lecture/demo/create-process/README.md:62]: Nice. This is what I'm talking about. But please use the same PIDs in the output of `pstree` and in the program output. Make manual changes if required.

File: content/chapters/compute/lecture/demo/create-process/README.md (hunk 7 above)
Comment (razvand) [reply content/chapters/compute/lecture/demo/create-process/README.md:56]: I say we keep things simple. As I said in [the corresponding PR](https://github.com/teodutu/operating-systems/pull/5), @Alex-deVis 's approach is much too complicated. It makes it hard for the reader to understand what the program is supposed to do.

**Review by razvand — APPROVED** [review]
Hi, @teodutu . I'm tentatively approving this. See my minor comments. After correcting them, we can merge this.

**Pull request merged**